### PR TITLE
Stabilize TryFrom / TryInto, and tweak impls for integers

### LINKED ISSUE #33417: Tracking issue for TryFrom/TryInto traits
Tracking issue for https://github.com/rust-lang/rfcs/pull/1542

----

To do:

* [ ] Is existing documentation satisfactory?
* [x] https://github.com/rust-lang/rust/pull/56796 Change bounds on `TryFrom` blanket impl to use `Into` instead of `From`
* [ ] (Re)stabilization PR

### PR DESCRIPTION
**Pull request by SimonSapin**
Fixes https://github.com/rust-lang/rust/issues/33417 (tracking issue)

----

This adds:

* `impl From<u16> for usize`
* `impl From<i16> for isize`
* `impl From<u8> for isize`

… replacing corresponding `TryFrom<Error=!>` impls. (`TryFrom` still applies through the generic `impl<T, U> TryFrom<U> for T where T: From<U>`.) Their infallibility is supported by the C99 standard which (indirectly) requires pointers to be at least 16 bits.

The remaining `TryFrom` impls that define `type Error = !` all involve `usize` or `isize`. This PR changes them to use `TryFromIntError` instead, since having a return type change based on the target is a portability hazard.

Note: if we make similar assumptions about the *maximum* bit size of pointers (for all targets Rust will ever run on in the future), we could have similar `From` impls converting pointer-sized integers to large fixed-size integers. RISC-V considers the possibility of a 128-bit address space (RV128), which would leave only `impl From<usize> for u128` and `impl From<isize> for u128`. I [found](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/20171017a-cheri-poster.pdf) some [things](http://www.csl.sri.com/users/neumann/2012resolve-cheri.pdf) about 256-bit “capabilities”, but I don’t know how relevant that would be to Rust’s `usize` and `isize` types.

I chose conservatively to make no assumption about the future there. Users making their portability decisions and using something like `.try_into().unwrap()`.

----

Since this feature already went through FCP in the tracking issue https://github.com/rust-lang/rust/issues/33417, this PR also proposes **stabilize** the following items:

* The `convert::TryFrom` trait
* The `convert::TryFrom` trait
* `impl<T> TryFrom<&[T]> for &[T; $N]` (for `$N` up to 32)
* `impl<T> TryFrom<&mut [T]> for &mut [T; $N]` (for `$N` up to 32)
* The `array::TryFromSliceError` struct, with impls of `Debug`, `Copy`, `Clone`, and `Error`
* `impl TryFrom<u32> for char`
* The `char::CharTryFromError` struct, with impls of `Copy`, `Clone`, `Debug`, `PartialEq`, `Eq`, `Display`, and `Error`
* Impls of `TryFrom` for all (?) combinations of primitive integer types where `From` isn’t implemented.
* The `num::TryFromIntError` struct, with impls of `Debug`, `Copy`, `Clone`, `Display`, `From<!>`, and `Error`

Some minor remaining questions that I hope can be resolved in this PR:

* Should the impls for error types be unified?
* ~Should `TryFrom` and `TryInto` be in the prelude? `From` and `Into` are.~ (Yes.)

### CODE REVIEW

**Comment by SimonSapin** [comment]
CC @rust-lang/libs 

r? @sfackler

**Comment by SimonSapin** [comment]
> Should TryFrom and TryInto be in the prelude? From and Into are.

I think they should, so I added a commit that makes them so. Let’s see if anyone objects :)

**Comment by SimonSapin** [comment]
We already did the FCP dance in the tracking issue, but just to get eyes on this again:

@rfcbot fcp merge

**Comment by ollie27** [comment]
>The remaining TryFrom impls that define type Error = ! all involve usize or isize. This PR changes them to use TryFromIntError instead, since having a return type change based on the target is a portability hazard.

The problem with this is that when (and I hope it's a when) the missing `From` impls for `usize` and `isize` are added then it will be a breaking change as the error type will have to change from `TryFromIntError` to `!`. It also doesn't actually fix the portability hazard when we consider the C int type defs which were one of the main motivating examples for these traits. For example the error type for `TryFrom<c_long> for i32` depends on the platform. If that is too much of a portability hazard then these traits are blocked on the portability lint (#41619).

**Comment by SimonSapin** [comment]
@ollie27 Which impls are you saying are missing? After this PR I believe that that every impl that *can* be implemented in on all platforms between primitive integers are already implemented. There shouldn’t be any we haven’t gotten around to *yet*.

As to type aliases like `c_long`, I think this problem is inherent to using a type alias whose definition changes across platform. However if you’re aware of this issue it is possible to write portable code with `TryFrom<c_long>`: in a function or method that returns `Result<_, TryFromIntError>`, use something like `x.try_into()?`. This works out because `TryFromIntError` implements both `From<TryFromIntError>` and `From<!>`.

**Comment by ollie27** [comment]
>Which impls are you saying are missing? 

All of the impls that aren't strictly speaking portable. They were proposed in #37423 but postponed for the portability lint.

> However if you’re aware of this issue it is possible to write portable code

Exactly. I don't see why people can't do this for `usize` and `isize`.

**Comment by SimonSapin** [comment]
As far as I can tell the portability lint wouldn’t help much here. We want `TryFrom` to always be available between any two integer types, the only thing that might change is the associated `Error` types. But the portability lint can only lint against using some items at all, it cannot be used to opt into *changing* an associated type.

> I don't see why people can't do this for usize and isize.

Just because a work around exists doesn’t mean it’s not bad. This PR tries to avoid this situation where we can.

**Comment by ollie27** [comment]
Just to be clear, are you saying that the impls proposed in #37423 will never be added? If you are then I strongly oppose that. Having `From` impls to statically assert that conversions can never fail is far more important than the portability hazard of error types being `TryFromIntError` or `!` depending on the platform.

**Comment by SimonSapin** [comment]
With this PR as originally proposed, indeed those could never be added as they would conflict with impls being stabilized here.

When discussing this offline @sfackler suggested that with the portability lint we could have some `From` impls conditionally compiled based on `#[cfg(target_pointer_width = …)]` and linted as non-portable, and `TryFrom<Error=TryFromIntError>` impls with opposite `cfg`, also linted as non-portable. This could work, but with the downside that could that uses these impls but would in fact be portable (because it either doesn’t care about the error type or does something with it like `?`’s conversion that works in both cases) is still required to "opt into non-portability" to silence the lint. Of course this also requires the portability lint to be implemented, and to work even when non-portable impls are used indirectly, "through" a generic impl like `TryFrom<T> for U where U: From<T>` or `Into<U> for T where U: From<T>`.

The set of impls that would be affected is not obvious so I tried to make an exhaustive table (with the same assumptions about pointer width than in this PR’s original message: at least 16 bits, but no assumed upper bound)

<table>
<tr>
<th>Source →<br>Dest ↓</th>
<th>u8</th>
<th>u16</th>
<th>u32</th>
<th>u64</th>
<th>u128</th>
<th>usize</th>
<th>i8</th>
<th>i16</th>
<th>i32</th>
<th>i64</th>
<th>i128</th>
<th>isize</th>
</tr>
<tr>
<th>u8</th>
<td>!</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
</tr>
<tr>
<th>u16</th>
<td>!</td>
<td>!</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>NP</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
</tr>
<tr>
<th>u32</th>
<td>!</td>
<td>!</td>
<td>!</td>
<td>E</td>
<td>E</td>
<td>NP</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
</tr>
<tr>
<th>u64</th>
<td>!</td>
<td>!</td>
<td>!</td>
<td>!</td>
<td>E</td>
<td>NP</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
</tr>
<tr>
<th>u128</th>
<td>!</td>
<td>!</td>
<td>!</td>
<td>!</td>
<td>!</td>
<td>NP</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
</tr>
<tr>
<th>usize</th>
<td>!</td>
<td>!</td>
<td>NP</td>
<td>NP</td>
<td>NP</td>
<td>!</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
</tr>
<tr>
<th>i8</th>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>!</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
</tr>
<tr>
<th>i16</th>
<td>!</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>!</td>
<td>!</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>NP</td>
</tr>
<tr>
<th>i32</th>
<td>!</td>
<td>!</td>
<td>E</td>
<td>E</td>
<td>E</td>
<td>NP</td>
<td>!</td>
<td>!</td>
<td>!</td>
<td>E</td>
<td>E</td>
<td>NP</td>
</tr>
<tr>
<th>i64</th>
<td>!</td>
<td>!</td>
<td>!</td>
<td>E</td>
<td>E</td>
<td>NP</td>
<td>!</td>
<td>!</td>
<td>!</td>
<td>!</td>
<td>E</td>
<td>NP</td>
</tr>
<tr>
<th>i128</th>
<td>!</td>
<td>!</td>
<td>!</td>
<td>!</td>
<td>E</td>
<td>NP</td>
<td>!</td>
<td>!</td>
<td>!</td>
<td>!</td>
<td>!</td>
<td>NP</td>
</tr>
<tr>
<th>isize</th>
<td>!</td>
<td>NP</td>
<td>NP</td>
<td>NP</td>
<td>NP</td>
<td>E</td>
<td>!</td>
<td>!</td>
<td>NP</td>
<td>NP</td>
<td>NP</td>
<td>!</td>
</tr>
</table>

Legend:

* `!`: this conversion never fails can be implemented with `From` on any platform
* `E`: this conversion can fail and must be implemented with `TryFrom<Error=TryFromIntError>` on every platform
* `NP` assuming a functional portability lint, this conversion could on some platforms be a non-portable infallible `From` impl

Whatever else we do, the few `From` impls for `!` entries that are not already stable can be added now.

I think there are four things we can do with the rest:

1. This PR as originally proposed: stabilize with apparently-fallible `TryFrom<Error=TryFromIntError>` impls for all E and NP entries, on all platforms.
2. Wait until the portability lint is implemented and use it as described above
3. Stabilize the traits now, but *remove* impls corresponding to `NP` entries in the table. Ship without them. Later, decide to do either 1. or 2.
4. Same as 3. but since it’s non-obvious what combinations of types are `NP` in the table also remove all `TryFrom` impls where `usize` or `isize` is involved, because that’s easier to understand. (But keep the `!` entries since some of those `From` impls are already stable.)

I’m now leaning toward 3 or 4.

**Comment by sfackler** [comment]
I think 4 may be the cleanest but 3 also seems fine.

**Comment by SimonSapin** [comment]
I’ve implemented 3 since that seems more useful. Most users don’t need to figure out mentally which impls exist or not, they can just try to use it and see what happens.

I’ve had to modify range iterators and `io::Cursor` since they relied on some of the impls being removed.

r? @sfackler

**Comment by SimonSapin** [comment]
(I thought that `try` tested more things than it does. I cancelled those builds.)

File: src/libcore/tests/num/mod.rs [hunk 1]
```diff
@@ -319,41 +320,44 @@ assume_usize_width! {
...
-        #[cfg(target_pointer_width = "64")] {
-            test_impl_try_from_always_ok! { test_try_u16isize, u16, isize }
-            test_impl_try_from_always_ok! { test_try_u32usize, u32, usize }
-            test_impl_try_from_always_ok! { test_try_u32isize, u32, isize }
-            test_impl_try_from_always_ok! { test_try_i32isize, i32, isize }
-            test_impl_try_from_always_ok! { test_try_u64usize, u64, usize }
-            test_impl_try_from_always_ok! { test_try_i64isize, i64, isize }
-        }
-    );
+// FIXME: Some impls temporarily removed, per
```
Comment (sfackler) [inline src/libcore/tests/num/mod.rs]: We should probably clear this and the commented out impls out entirely.

**Comment by dgriffen** [comment]
I caught this too late but I think there is an issue with including `TryFrom` and `TryInto` in the prelude. Namely that it is a breaking change and actually broke the `http` crate on nightly. The compiler is actually very helpful and suggests new code to write to fix the issue but I think there should be some more consideration of this.

**Comment by SimonSapin** [comment]
Re being late: that’s ok, this is why we have release channels.

Could you give details of the breakage?

**Comment by dgriffen** [comment]
Sure, this is the type of error that is being produced:
```
error[E0034]: multiple applicable items in scope
   --> src\method.rs:325:9
    |
325 |         Method::try_from(t.as_bytes())
    |         ^^^^^^^^^^^^^^^^ multiple `try_from` found
    |
    = note: candidate #1 is defined in an impl of the trait `std::convert::TryFrom` for the type `_`
note: candidate #2 is defined in the trait `convert::HttpTryFrom`
   --> src\convert.rs:22:5
    |
22  |     fn try_from(t: T) -> Result<Self, Self::Error>;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = help: to disambiguate the method call, write `convert::HttpTryFrom::try_from(...)` instead
```

Looks like there was already a try_from method created in the `http` crate to get around the fact that `TryFrom` wasn't stable until now. This caused a name conflict since `TryFrom` is included in prelude. I doubt that `http` is the only crate that will run into this problem.

**Comment by SimonSapin** [comment]
I think that this kind of breakage is considered minor by https://rust-lang.github.io/rfcs/1105-api-evolution.html#minor-change-implementing-any-non-fundamental-trait since it is possible to write that `http` crate code in a way that works both before and after the change (`HttpTryFrom::try_from` or `<Method as HttpTryFrom>::try_from`). To avoid it entirely we’d need to never extend the API of existing standard library types.

That said we can do a Crater run to evaluate the impact in this case, and if we judge that necessary backport a removal from the prelude during the beta cycle (though I do think that would be unfortunate).

**Comment by Mark-Simulacrum** [comment]
We'll probably just look at beta Crater results once that's branched and decide based on that. I'm personally somewhat "annoyed" that people didn't change their APIs not to use `try_from` since it was an unstable, but likely to be stabilized feature. Perhaps we should have some warning about doing so for the future? I think @aidanhs has changed the resolution code recently to enable that, though I'm not quite certain...

**Comment by SimonSapin** [comment]
@Mark-Simulacrum re warning, you may be thinking of https://github.com/rust-lang/rust/pull/48552 though it (should) apply to traits in scope. In this case we should have added these traits to the prelude earlier, while they were unstable.

**Comment by withoutboats** [comment]
I'm surprised that prelude trait methods aren't "shadowed" the same way prelude items are shadowed.

**Comment by petrochenkov** [comment]
@withoutboats 
>I'm surprised that prelude trait methods aren't "shadowed" the same way prelude items are shadowed.

The decision happened in https://github.com/rust-lang/rust/pull/31908

**Comment by SimonSapin** [comment]
> We'll probably just look at beta Crater results once that's branched and decide based on that.

I agree, let’s wait for these results.

**Comment by bill-myers** [comment]
Couldn't this be fixed by ignoring prelude trait methods if they conflict with anything else?

**Comment by ltratt** [comment]
I appreciate that those of us who relied on unstable features get what we deserve, but before I start refactoring a load of code which used `try_from`, I thought I would check. Summary: I have code which uses `try_from` from usize to u32 so with this PR:

```
   --> src/lib/lexer.rs:255:18
    |
255 |             len: u32::try_from(len).unwrap(),
    |                  ^^^^^^^^^^^^^ the trait `std::convert::From<usize>` is not implemented for `u32`
    |
    = help: the following implementations were found:
              <u32 as std::convert::From<std::net::Ipv4Addr>>
              <u32 as std::convert::From<u8>>
              <u32 as std::convert::From<char>>
              <u32 as std::convert::From<u16>>
    = note: required because of the requirements on the impl of `std::convert::TryFrom<usize>` for `u32`
```

But, unfortunately, I don't think I can redefine try_from locally:

```rust
impl TryFrom<usize> for u32 {
    fn try_from(value: usize) -> Result<Self, Self::Error> {
        if value > Self::max_value() {
            Err(())
        }
        Ok(value as Self)
    }
}
```

because then I get:

```
error[E0119]: conflicting implementations of trait `std::convert::TryFrom<usize>` for type `u32`:
  --> src/lib/lexer.rs:41:1
   |
41 | impl TryFrom<usize> for u32 {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `core`:
           - impl<T, U> std::convert::TryFrom<U> for T
             where T: std::convert::From<U>;
   = note: upstream crates may add new impl of trait `std::convert::From<usize>` for type `u32` in future versions

error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
  --> src/lib/lexer.rs:41:1
   |
41 | impl TryFrom<usize> for u32 {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ impl doesn't use types inside crate
   |
   = note: the impl does not reference any types defined in this crate
   = note: define and implement a trait or new type instead
```

Am I missing out something, or do I have to add a new function outside of the int types and rewrite all calls to use it? (ccing @snim2 who first spotted this in https://github.com/softdevteam/lrlex).

**Comment by SimonSapin** [comment]
@ltratt This is one of the impls that were removed in this PR as discussed above. This was done to allow stabilization of the trait not to be blocked on picking one of two possible mutually-incompatible ways to have these impls. (One of which would itself be blocked on the portability lint being implemented.) I agree that not having this functionality at all for new sucks. Adding them back is tracked at https://github.com/rust-lang/rust/issues/49415.

(I’m personally in favor of having these impls as fallible `TryFrom<Error=TryFromIntError>` on all platforms, which we could do Very Soon® if we decide to give up on infallible `From` impl on only some platforms with a portability lint.)

**Comment by ltratt** [comment]
@SimonSapin Thanks for your reply -- much appreciated! I'd originally hoped that I could replace a few `::try_from()` calls with another function (which is a bit annoying, but not very difficult, to replace) but now I've realised that I use `TryFrom` as a trait constraint in several places (which is hard, and perhaps even impossible, to replace because we're forbidden from adding behaviour in our crate). We've (well, @snim2) temporarily solved this by rolling back our nightly a few days, although obviously that's going to catch up with us soon ;) I don't have any opinion on fallibility/infallibility myself, so I shall cross my fingers that a consensus is reached soon. Thanks again for your fast reply!

**Comment by SimonSapin** [comment]
One work-around ([used by this PR itself](https://github.com/rust-lang/rust/pull/49305/files#diff-e8903f372636cff6b22e471a663b7437R464) in another prat of std) is to have your own trait with a generic impl that forwards to `TryFrom`’s impls, plus additional impls that you define. However:

* It might not be acceptable if the generic bound is part of a public API
* It will break (because of conflicting impls) when https://github.com/rust-lang/rust/issues/49415 is fixed and these impls are added to `TryFrom` (one way or another). This could be avoided by not using `TryFrom` at all and having a completely separate trait.

**Comment by udoprog** [comment]
FYI, stdweb is also broken on nightly because TryFrom/TryInto is in the prelude, first error:

**Comment by TeXitoi** [comment]
rust-geojson also fails.

**Comment by ltratt** [comment]
@SimonSapin Indeed, we're sometimes using `TryFrom` in public APIs, which is why I think I'm in trouble.

**Comment by carllerche** [comment]
@SimonSapin 

The policy you refer includes:

> All that said, it is incumbent on library authors to ensure that such "minor" changes are in fact minor in practice: if a conflict like t.foo() is likely to arise at all often in downstream code, it would be advisable to explore a different choice of names. More guidelines for the standard library are given later on.

There was some discussion in https://github.com/hyperium/http/issues/192.

I will repeat my main point here:

The Rust team has advertised stability as a feature in such a way that it implied to me that pinning the stable rust version was not needed. To downstream users, it doesn't actually matter if a change is “minor” or not, it matters if the build will break. If the Rust team is not reasonably able to provide a guarantee of stability, that is fine but should be advertised and downstream users can adjust best practices by, for example, pinning the stable version of Rust.

Again if this change lands on Rust stable, it will break the existing frozen Conduit builds due to the fact that we did not pin the version of Rust but dependencies are pinned. We cannot fix these past version releases as they have been cut.

**Comment by carllerche** [comment]
To be explicit, I understand that almost literally any change to `std` **could** be a breaking change. My understanding of the policy, especially given how the snippet I quoted above reads, is that if a change turns out to be breaking **in practice** then it is a non starter ("minor" or not), and that a *best effort* would be made to changes that break existing code.

If this interpretation is not correct, I would appreciate it if the Rust team goes on the record to clarify the policy, because at least I and probably others are under the assumption that we can expect stability across Rust releases. If this assumption is not correct, we need to change our practices to start pinning the compiler version in applications.

**Comment by seanmonstar** [comment]
With the new 2018 edition coming soon, it seems reasonable to include these in the prelude only in the newer editions. It prevents any breakage.

**Comment by SimonSapin** [comment]
@carllerche Completely agree. We have roughly 6 weeks before this change reaches the stable channel. 1.26 goes to beta this week (today?) and the plan was to wait for crater results for that beta to get an idea of how bad the breakage is in the ecosystem. I’ll talk with other libs team member today to see what they think of reverting the prelude addition sooner than that based on the reports on `http` and other crates in this thread.

@seanmonstar Indeed we’ve discussed having a larger prelude based on a crate’s edition. That could be an opportunity to add `TryFrom` and `TryInto` without breaking existing code.

**Comment by tirkarthi** [comment]
To add to this `rand` also fails on nightly.

Build : https://travis-ci.org/rust-lang-nursery/rand/jobs/359738013

**Comment by tirkarthi** [comment]
I downloaded the modules metadata from https://github.com/rust-lang/crates.io-index and inserted them into MongoDB to analyze the number of modules that might break. I did a query for stdweb and http. I also did a query on the number of packages that depend on rand which uses stdweb. The approx results are as below : 

stdweb - 15 dependants
http - 20 dependants
rand - 1070 dependants

Gist of dependencies : https://gist.github.com/tirkarthi/846f9d2fd87c90463c5ca9cf99f3d44d
File that can be used to do queries locally : https://gist.github.com/tirkarthi/846f9d2fd87c90463c5ca9cf99f3d44d#file-scraper-py

**Comment by frewsxcv** [comment]
FYI, @SimonSapin opened a PR which reverts the `prelude` additions: https://github.com/rust-lang/rust/pull/49518

**Comment by SimonSapin** [comment]
Oops, I meant to comment about this here but forgot. Thanks @frewsxcv.

As mentioned in that PR I’d still like to figure something out later, but for now let’s fix immediate breakage.

**Comment by SimonSapin** [comment]
:bell: :bell: **Proposed resolution** for https://github.com/rust-lang/rust/issues/49415:

https://github.com/rust-lang/rust/pull/51564 restores `TryFrom` impls that are infallible on some platforms only as **always fallible**, on all platforms, at the type system level (with `Error=TryFromIntError`).

**Comment by SoniEx2** [comment]
How does one handle something like libc::time_t using the new TryFrom?

**Comment by SimonSapin** [comment]
@SoniEx2 A merged or otherwise closed PR/issue is often not a good place to ask question. Consider asking on https://github.com/rust-lang/rust/issues/33417 or on https://internals.rust-lang.org/ instead. But also: what do you mean by "handle"? Convert to/from what other type?

**Comment by SoniEx2** [comment]
say I need to turn a time_t to/from u64. since time_t can change types, this seems like I'd have to be putting either "as" or #[cfg] everywhere? none of them let me write code that "just works"...

I wonder why we couldn't have multiple otherwise identical TryFroms with different error types...

**Comment by SimonSapin** [comment]
It looks like `time_t` is always either `i32` or `i64`. So converting to `u64` is fallible in both cases because of negative values. Converting from `u64` is also fallible in both cases because of very large values. So you’d always get `TryFrom` impls with `type Error = TryFromIntError;`.

Note that infallible conversions still get a `TryFrom` impl through `impl<T, U> TryFrom<U> for T where T: From<U> { type Error = !; }`. So if you were to convert `time_t` to/from `i64` instead of `u64` for example, then the `try_*` methods would be available on all platforms but with a varying error type. However it’s still possible to write portable code without using `#[cfg]`. For example:

```rust
fn foo(x: u64) -> Result<time_t, TryFromIntError> {
    Ok(x.try_into()?)
}
```

This works because `?` does error type conversion. Depending on the platform this ends up calling `impl From<!> for TryFromIntError` or the no-op `impl<T> From<T> for T`.

**Pull request merged**